xclim/indices/generic.py

```diff
@@ -897,3 +898,103 @@ def first_day_threshold_reached(
     )
     out.attrs.update(units="", is_dayofyear=np.int32(1), calendar=get_calendar(data))
     return out
+
+
+def _get_zone_bins(
+    zone_min: Quantified | None,
+    zone_max: Quantified | None,
+    zone_step: Quantified | None,
+):
+    units = pint2cfunits(str2pint(zone_step))
+    mn, mx, step = (
+        convert_units_to(str2pint(z), units) for z in [zone_min, zone_max, zone_step]
+    )
+    bins = np.arange(mn, mx + step, step)
+    if (mx - mn) % step != 0:
+        warnings.warn(
+            "`zone_max` - `zone_min` is not an integer multiple of `zone_step`. Last zone will be smaller."
+        )
+        bins[-1] = mx
+    return xr.DataArray(bins, attrs={"units": units})
+
+
+def get_zones(
+    da: xr.DataArray,
+    zone_min: Quantity | None = None,
+    zone_max: Quantity | None = None,
+    zone_step: Quantity | None = None,
+    zone_1: Quantity | None = None,
+    bins: xr.DataArray | list[Quantity] | None = None,
+    exclude_boundary_zones: bool = True,
+    close_last_zone_right_boundary: bool = True,
+) -> xr.DataArray:
+    r"""Divide data into zones and attribute a zone coordinate to each input value.
+
+    Divide values into zones corresponding to bins of width zone_step beginning at zone_min and ending at zone_max.
+    Bins are inclusive on the left values and exclusive on the right values.
+
+    Parameters
+    ----------
+    da : xarray.DataArray
+        Input data
+    zone_min : Quantity | None
+        Left boundary of the first zone
+    zone_max : Quantity | None
+        Right boundary of the last zone
+    zone_step: Quantity | None
+        Size of zones
+    zone_1 : Quantity | None
```

Comment:
The zone_1 thing is not super clear to me.

Reply:
Take for example USDA method
in this case, `bins = [-60, -55, -50, ... 70]`

`[-60 degF, -55 degF[` is the first bin. By default, `np.digitize(-58, bins) ==1`
`[-55 degF, -50 degF[` is the second bin. `np.digitize(-52, bins) ==2`

This was an option to let choose which bin is labelled as "1". I think there were some dicussions allowing lower temperature at some point. This would let for instance extend the USDA zones to lower temperatures while keeping the original ordering. One would set the first bin to [-60 degF, -55 degF[. A lower bin [-65, -60] would then have index "0". THe lowet one "-1", etc.

This could be taken care of at higher level if it's better.

Reply:
What throws me off here is that zone_1 is a Quantity. What you're describing feels more like an integer offset.

Reply:
Yes, that's what I had in mind by higher-level treatment, just take zones -> zones + zone_offset. 

Implementing this in the function would just be `return zones + zone_offset` instead of `return zones`. Is  it still worth keeping it?

Reply:
Since this logic is not user facing, I'm not sure the added complexity is worth it. 
You could keep the function private if we're not sure how exactly how it should be used and what signature it should have.